### DISCUSSION TITLE
Our builds, base image and such

### DISCUSSION THREAD

**Discussion by haugene** [haugene/docker-transmission-openvpn#2595] [General]
This is kind of a follow up to https://github.com/haugene/docker-transmission-openvpn/issues/2469, https://github.com/haugene/docker-transmission-openvpn/pull/2590 and https://github.com/haugene/docker-transmission-openvpn/pull/2589 where we've been touching on issues like the choice of base image as well as build optimizations and issues. Even if this ends up as a conversation with myself, or just me and @pkishino I think it's good to have it in an open thread for transparency and later reference.

Short summaries of the past, present and future:

**Past:** We were Ubuntu based, then had a period of dual builds Ubuntu/Alpine before choosing to go with Alpine because of smaller image size and wanting to avoid supporting two versions in parallel. After a couple of years with that we went back to Ubuntu because we wanted stability instead of fast update of underlying packages and libraries. One example issue is #1861 

**Present:** We went back to Ubuntu which meant we also un-intentionally downgraded Transmission to 2.94 from 3.00. This caused some issues but we stayed the course. In Ubuntu 22.04 we're back at 3.00 but that seems to cause memory leaks #2469. And now from version 5.0.0 of the image we're running Transmission 4.x anyways and we build it from source as part of the image build so we're not depending on support from the apt repositories.

**Future:** Right now we're not getting the best of any world I think. We're having memory issues on 22.04 and we have really slow build times due to us doing the Transmission builds ourselves. Alpine has a strong card on its hand with already having Transmission 4.0.3 in the `edge/community` repo. I played around with it a bit and that's not straight forward either without basing the whole image on alpine:edge which probably will take us back to instability as we experienced before.

I'm open for suggestions. But as we weigh our options I'm inclined to do a minor release 5.1 where we downgrade base image to ubuntu:20.04, and introduce caching in our docker builds to speed things up. Hopefully that makes it stable again and speeds up the average build time enough to live with. We won't downgrade Transmission so I think we're fine, people don't care too much about the version of openvpn, curl or any other library in the container.

Thoughts?

**Comment by pkishino** [comment]
Not sure there is any point downgrading to 22.04 when the leaks afaik were related to transmission 3.0, no?

**Reply by haugene** [reply]
Definitely. Maybe that closes the discussion already :sweat_smile:
I see you've asked for confirmation in #2469 so let's see if this is no longer present for release 5.0.2 and then we'll just stay on 22.04 for sure :+1:

**Comment by ksurl** [comment]
Agreed, the current build times for new releases are really slow. My proposal is to move the TransmissionBuilder stage in the Dockerfile to it's own image that updates when transmission releases new versions. Decoupling this will speed up build times for the image release.

You can automate updating the build image with an action that checks for new versions and auto bumps the image version (it would make the most sense to make the tags match the transmission version for simplicity) with renovate bot (supports github releases to check for new transmission versions https://docs.renovatebot.com/modules/datasource/github-releases/) when a new version is detected.

So new builds of transmission only run when there's a new release. New image releases for this repo will be independent and only be installing dependencies, web ui's and adding the other scripts and components.

the build image would be something haugene/transmission-build with it's own workflow to detect the new transmission releases and run independently of the image build for haugene/transmission-openvpn

**Reply by ksurl** [reply]
this was prompted by checking on the new build for v5.1.0 and it was still running an hour after the release showed in my github notifications.

**Reply by ksurl** [reply]
might be simpler to setup a workflow to check for releases and then build a new image if found using a bash script with `jq` than setting up full blown renovate (unless you want to manage other dependencies in the repo as well)

**Reply by ksurl** [reply]
please take a look at https://github.com/haugene/docker-transmission-openvpn/pull/2691

**Reply by pkishino** [reply]
Thanks

**Reply by ksurl** [reply]
It looks like the latest fixes on the main build workflow is good (it builds all 3 architectures for the rss and proxy), except for the main image since the haugene/transmission-builder image doesn't exist yet. Is there a way to manually trigger the builder workflow?

The version check workflow is set to run hourly so if you wanted to downgrade upstream/transmission-version.txt to 4.0.3 (which should trigger a build), it should bump it 4.0.4 again and trigger another new build. probably should have put them in 2 different commits so it could have the new workflow before it starts checking for the version file

**Reply by pkishino** [reply]
this should be it ..
https://github.com/haugene/docker-transmission-openvpn/actions/workflows/docker-transmission-builds.yml

**Reply by ksurl** [reply]
right, it failed due to the file typo which has been fixed. are you able to re-run it so it uses the latest Dockerfile with the fix? I don't have permissions to the actions on the repo other than viewing the status/output.

**Comment by pkishino** [comment]
ok, added a workflow_dispatch to the job and triggered it..

**Reply by ksurl** [reply]
looks like amd64 finished and arm/64 is still running. should be good on the next real run for the main image build (really fast).

**Reply by pkishino** [reply]
hmm, so, somehow the sha5 doesn't match when trying to build the final container image for amd64 version..

**Reply by pkishino** [reply]
so somehow the manifest isn't found for the platform..strange

**Reply by ksurl** [reply]
it looks like the builder image only pushed the arm build. maybe best to go back to the single workflow (explicitly setting platforms in the action) instead of matrix builds since that was more reliable. regardless, having the builder image will speed up the release builds. I just tested matrix builds on my own test repo and had the same issue (only pushed arm builds). 

or if we want to keep matrix builds, we could try separating the build cache as seen here:

https://github.com/docker/build-push-action/issues/846#issuecomment-1542958636

**Reply by ksurl** [reply]
I tested that out and nothing worked. still only pushes 1 architecture. I think we should go back to single build for platforms but keep the matrix to have a single workflow for the 3 image builds. and also 1 build for the builder image. must be something with the cache and the simultaneous build push. the sha doesn't match since they're not in the same builder probably.

I'll update the PR's

**Reply by ksurl** [reply]
see #2704 and #2705

**Reply by edgd1er** [reply]
Another idea would be to build and commit v4 deb file. Not very neat but efficient. Image build time with buildx woulld be much quicker.
Docker has an option to output directories.

**Reply by ksurl** [reply]
oh yeah, could do artifact uploads as releases but that requires a separate repo to not clutter the main releases. committing non text files is not very efficient (it will essentially store each copy). I think the current solution works for now.

**Reply by ksurl** [reply]
I think this one is the old workflow with matrix build? https://github.com/haugene/docker-transmission-openvpn/actions/runs/6078202126

may not be needed

**Reply by ksurl** [reply]
looks like they're both canceled.

**Reply by pkishino** [reply]
Yeah, thanks, must have accidentally cancelled both

**Reply by ksurl** [reply]
you re-ran the one with the matrix platform build and it only pushed the arm build

**Comment by edgd1er** [comment]
When using separate builds , manifest has to be edited manually.
Docker buildx = easy way, docker manifest =hard way.
https://www.docker.com/blog/multi-arch-build-and-images-the-simple-way/

**Reply by ksurl** [reply]
thanks, we were using buildx with the docker-build-push action. it seems building each platform separately with matrix multi-runner breaks the manifest.

**Reply by pkishino** [reply]
current issue is related to this.. 
https://github.com/actions/runner/issues/1923

**Reply by ksurl** [reply]
ah I see it's a different issue (github related). may need to manually retrigger transmission build workflow as well since the latest runs that did get past the checkout step were unable to find the image for 4.0.4 on amd64, arm64 since only arm was pushed.

**Reply by pkishino** [reply]
Thanks for the help!! Back on track now

**Reply by ksurl** [reply]
> Thanks for the help!! Back on track now

Not quite yet. There is a little cleanup to do since the 5.1.1 pre release failed build and push now replaced 5 and 5.1 tags with only the arm or arm64 image on docker hub for rss, proxy and the main image.

It needs a repush on either 5.1.0 or 5.1.1 or maybe 5.1.2 as a clean new version.

**Comment by ksurl** [comment]
Wow what a difference on the new build time using the prebuilt transmission builder. 6-16min down from 3+ hours. 

For future version updates, I think all you need to do is change the image version in the main Dockerfile for haugene/transmission-builder.

The other 2 workflows should automatically detect and build new images when transmission has an upstream release. I'll keep an eye on it when they do another release though that seems to be random.
